// create-react-web-component/dist/reactComponent/CustomComponent.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("@webcomponents/webcomponentsjs/webcomponents-bundle.js");
require("@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js");
var react_1 = __importDefault(require("react"));
var react_dom_1 = __importDefault(require("react-dom"));
var react_shadow_1 = require("react-shadow");
var EventContext_1 = require("../components/EventContext");
var componentAttributes;
var componentProperties;
var rootComponent;
var shadow;
exports.setComponentAttributes = function (attributes) {
    componentAttributes = attributes;
};
exports.setComponentProperties = function (properties) {
    componentProperties = properties;
};
exports.setRootComponent = function (component) {
    rootComponent = component;
};
exports.setMode = function (shadowOption) {
    shadow = shadowOption;
};
var CustomComponent = /** @class */ (function (_super) {
    __extends(CustomComponent, _super);
    function CustomComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.eventDispatcher = function (event) {
            _this.dispatchEvent(event);
        };
        return _this;
    }
    Object.defineProperty(CustomComponent, "observedAttributes", {
        get: function () {
            return Object.keys(componentAttributes).map(function (k) { return k.toLowerCase(); });
        },
        enumerable: true,
        configurable: true
    });
    CustomComponent.prototype.reactProps = function () {
        var _this = this;
        var attributes = {};
        Object.keys(componentAttributes).forEach(function (key) {
            attributes[key] = _this.getAttribute(key) || componentAttributes[key];
        });
        return __assign(__assign({}, attributes), componentProperties);
    };
    CustomComponent.prototype.connectedCallback = function () {
        this.mountReactApp();
    };
    CustomComponent.prototype.attributeChangedCallback = function (name, oldValue, newValue) {
        if (oldValue === newValue) {
            return;
        }
        this.mountReactApp();
    };
    CustomComponent.prototype.reactPropsChangedCallback = function (name, oldValue, newValue) {
        if (oldValue === newValue) {
            return;
        }
        componentProperties[name] = newValue;
        this.mountReactApp();
    };
    CustomComponent.prototype.disconnectedCallback = function () {
        react_dom_1.default.unmountComponentAtNode(this);
    };
    CustomComponent.prototype.mountReactApp = function () {
        var application = (react_1.default.createElement(EventContext_1.EventProvider, { value: this.eventDispatcher }, react_1.default.createElement(rootComponent, this.reactProps())));
        if (shadow !== undefined && !shadow) {
            react_dom_1.default.render(application, this);
        }
        else {
            var root = react_shadow_1.createProxy({ div: undefined });
            react_dom_1.default.render(react_1.default.createElement(root.div, null, application), this);
        }
    };
    return CustomComponent;
}(HTMLElement));
exports.default = CustomComponent;
